src/VisualStudio/Core/Def/ProjectSystem/OpenTextBufferProvider.cs

```diff
@@ -0,0 +1,313 @@
+            runningDocumentTable.AdviseRunningDocTableEvents(this, out _runningDocumentTableEventsCookie);
+
+            // We also need to check for any documents that might have been open before we subscribed. That we do have to do on the UI thread.
+            var listener = listenerProvider.GetListener(FeatureAttribute.Workspace);
+            var asyncToken = listener.BeginAsyncOperation(nameof(CheckForExistingOpenDocumentsAsync));
+            CheckForExistingOpenDocumentsAsync(threadingContext).CompletesAsyncOperation(asyncToken);
+        }
+
+        private void RaiseEventForEachListener(Action<IOpenTextBufferEventListener> action)
+        {
+            _foregroundAffinitization.AssertIsForeground();
+
+            foreach (var listener in _listeners)
+            {
+                try
+                {
+                    action(listener);
+                }
+                catch (Exception e) when (FatalError.ReportAndCatch(e, ErrorSeverity.Critical))
+                {
+                    // We'll catch the exception; this way if one listener is broken, we don't end up breaking other features that might no longer get events. Any exceptions would get caught by the
+                    // RunningDocumentTable itself which wouldn't report them in a useful way regardless.
+                }
+            }
+        }
+
+        private async Task CheckForExistingOpenDocumentsAsync(IThreadingContext threadingContext)
+        {
+            await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            foreach (var (filePath, textBuffer, hierarchy) in EnumerateDocumentSet())
+            {
+                // We might or might not have seen this file be opened if it was opened between when we subscribed to the running document table and when
+                // we got scheduled to the UI thread.
+                if (!_monikerToTextBufferMap.ContainsKey(filePath))
+                {
+                    _monikerToTextBufferMap = _monikerToTextBufferMap.Add(filePath, textBuffer);
+
+                    foreach (var listener in _listeners)
+                        listener.OnOpenDocument(filePath, textBuffer, hierarchy);
+                }
+            }
+        }
+
+        public void AddListener(IOpenTextBufferEventListener listener) => ImmutableInterlocked.Update(ref _listeners, static (array, listener) => array.Add(listener), listener);
+        public void RemoveListener(IOpenTextBufferEventListener listener) => ImmutableInterlocked.Update(ref _listeners, static (array, listener) => array.Remove(listener), listener);
+
+        public int OnAfterFirstDocumentLock(uint docCookie, uint dwRDTLockType, uint dwReadLocksRemaining, uint dwEditLocksRemaining)
+            => VSConstants.E_NOTIMPL;
+
+        public int OnBeforeLastDocumentUnlock(uint docCookie, uint dwRDTLockType, uint dwReadLocksRemaining, uint dwEditLocksRemaining)
+        {
+            if (dwReadLocksRemaining + dwEditLocksRemaining == 0)
```

Comment:
is this a check to verify that there's nothing else actually keeping this open?

Reply:
I guess? This check we've done since more or less the beginning of time cargo-cult style. :smile:

Reply:
fair!